### PR TITLE
Moved the failed login attemps to the DB 

### PR DESCRIPTION
**Pull request by michalpiatek**
(can be shared between the se services without affinity token now)

### CODE REVIEW

File: src/main/resources/db/migration/mysql/V1.6.1__add-failed-attempt-counter.sql [hunk 1]
```diff
@@ -0,0 +1 @@
+ALTER TABLE `identity` ADD `failed_login_attempts` INT NOT NULL DEFAULT 0;
```
Comment (charlesluokb) [inline src/main/resources/db/migration/mysql/V1.6.1__add-failed-attempt-counter.sql]: can we rename the file to `V1.6.1__add-failed-login-attempt-counter`

File: src/main/java/uk/gov/cshr/domain/Identity.java [hunk 2]
```diff
@@ -41,7 +44,7 @@
     public Identity() {
     }
 
-    public Identity(String uid, String email, String password, boolean active, boolean locked, Set<Role> roles, Instant lastLoggedIn, boolean deletionNotificationSent) {
+    public Identity(String uid, String email, String password, boolean active, boolean locked, Set<Role> roles, Instant lastLoggedIn, boolean deletionNotificationSent, Long failedLoginAttempts) {
```
Comment (charlesluokb) [inline src/main/java/uk/gov/cshr/domain/Identity.java]: if we have a bit time, I would either introduce a IdentityBuilder, or enable Lombok's annotation e.g @data

File: src/test/java/uk/gov/cshr/service/security/LoginAttemptServiceTest.java [hunk 3]
```diff
@@ -1,79 +1,82 @@
...
-        loginAttemptCache.put(email, 3);
-
-        when(identityService.existsByEmail(email)).thenReturn(true);
+        when(identityRepository.findFirstByActiveTrueAndEmailEquals(email)).thenReturn(identity);
+        when(identityRepository.findFirstByEmailEquals(email)).thenReturn(identity);
+        when(identityRepository.existsByEmail(email)).thenReturn(true);
 
         try {
             loginAttemptService.loginFailed(email);
         } catch (AuthenticationException e) {
```
Comment (charlesluokb) [inline src/test/java/uk/gov/cshr/service/security/LoginAttemptServiceTest.java]: can we add a new statement after line 74
`fail("AuthenticationException was not thrown")`

Comment (michalpiatek) [reply src/test/java/uk/gov/cshr/service/security/LoginAttemptServiceTest.java]: will add

File: src/main/java/uk/gov/cshr/domain/Identity.java (hunk 2 above)
Comment (michalpiatek) [reply src/main/java/uk/gov/cshr/domain/Identity.java]: Not a fan of lombok ;)

will provide a Builder with ObjectMother for tests

File: src/main/resources/db/migration/mysql/V1.6.1__add-failed-attempt-counter.sql (hunk 1 above)
Comment (michalpiatek) [reply src/main/resources/db/migration/mysql/V1.6.1__add-failed-attempt-counter.sql]: sure

File: src/test/java/uk/gov/cshr/data/provider/IdentityMother.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
+package uk.gov.cshr.data.provider;
+
+import org.springframework.stereotype.Service;
+import uk.gov.cshr.domain.Identity;
+
+import java.time.Instant;
+import java.util.HashSet;
+
+@Service
+public class IdentityMother {
```
Comment (charlesluokb) [inline src/test/java/uk/gov/cshr/data/provider/IdentityMother.java]: 👍  (looking forward to pair it with IdentityFather ;-) )

File: src/test/java/uk/gov/cshr/data/provider/IdentityMother.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+    public Identity provideIdentity(String uid, String email, String password) {
```
Comment (charlesluokb) [inline src/test/java/uk/gov/cshr/data/provider/IdentityMother.java]: could rename it to `giveIdentityBirth` ;-)

**Pull request closed without merging**